keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,438 @@
+- Cross-family connectivity: IPv4-to-IPv6 and IPv6-to-IPv4 connectivity is considered outside of the scope of this proposal, except in the case of ingress access, where cross-family access may be supported depending upon the chosen ingress controller.
+- CNI network plugins: Plugins other than the Bridge, PTP, and Host-Local IPAM plugins should support Kubernetes dual stack, but the development and testing of dual stack support for these other plugins is considered outside of the scope of this proposal.
+- Multiple IPs vs. Dual-Stack: Code changes will be done in a way to facilitate future expansion to more general multiple-IPs-per-pod and multiple-IPs-per-node support. However, this initial release will impose "dual-stack-centric" IP address limits as follows:
+  - Pod addresses: 1 IPv4 address and 1 IPv6 addresses per pod maximum
+  - Node addresses: 1 IPv4 address and 1 IPv6 addresses per pod maximum
+  - Service addresses: 1 IP address pod service
+- For simplicity, the [Kubernetes Endpoints API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.10/#endpoints-v1-core) will not be modified. Each IPv4/IPv6 address that is assigned to a backend pod will be treated and displayed as a separate endpoint for the corresponding service. Service load balancing (e.g. using kube-proxy) will be performed on a per-address basis, rather than being done on a per-dual-stack-endpoint basis.
+- Kube-DNS is expected to be End-of-Life soon, so dual-stack testing will be performed using coreDNS.
+- External load balancers that rely on Kubernetes services for load balancing functionality will only work with the IP family that matches the IP family of the cluster's service CIDR.
+- Dual-stack support for Kubernetes orchestration tools other than kubeadm (e.g. miniKube, KubeSpray, etc.) are considered outside of the scope of this proposal.
+
+## Proposal
+
+In order to support dual-stack in Kubernetes clusters, Kubernetes needs to have awareness of and support dual-stack addresses for pods and nodes. Here is a summary of the proposal (details follow in subsequent sections):
+
+- Kubernetes needs to be made aware of multiple IPs per pod (limited to one IPv4 and one IPv6 address per pod maximum).
+- Link Local Addresses (LLAs) on a pod will remain implicit (Kubernetes will not display nor track these addresses).
+- Backend pods for a service can be dual stack. For the first release of dual-stack support, each IPv4/IPv6 address of a backend pod will be treated as a separate Kubernetes endpoint.
+- Kube-proxy needs to drive iptables and ip6tables in parallel. This is required in order to allow exposing services via both IPv4 and IPv6, e.g. using Kubernetes:
+  - NodePort
+  - ExternalIPs
+- "Nice to Have" or stretch goal: IPVS proxying needs to support IPv4 and IPv6 services in parallel. However, this support depends upon resolution of [cloudnativelabs/kube-router Issue #307](https://github.com/cloudnativelabs/kube-router/issues/307).
```

Comment:
May want to split this into two items. One is IPVS via kube-router (as a replacement for kube-proxy). That depends on issue 307 for IPv6 support. The other is kube-proxy running in IPVS mode, which should work, provided the mentioned kube-proxy changes in this KEP are performed.

Reply:
@pmichali - Do you know which IPVS mode is (eventually) going to be the reference mode, IPVS via kube-router or IPVS via kube-proxy?

Reply:
I don't know which method is planned for the reference in the future. Maybe someone in the community has a better handle on that.

For kube-proxy, it appears that the default is still to use iptables currently.  I think kube-router is pushing for GA, but I don't know the timing or the plans for either of these two.

I have to admit, kube-router is a much easier way to set things up.

Reply:
I think we should focus on ipvs mode here, because it's part of core k/k

Reply:
@feiskyer - Sounds good, especially if you can help us out with getting IPVS mode working (V6-only first, then eventually dual-stack).